### PR TITLE
Abeeway

### PR DESCRIPTION
**Pull request by J2thatsme**
Will add more topics later, tests work.

### CODE REVIEW

File: types/abeeway/badgeTracker/acknowledge.schema.json [hunk 1]
```diff
@@ -0,0 +1,13 @@
+{
+  "$id": "https://akenza.io/abeeway/compactTracker/acknowledge.schema.json",
```
Comment (mariusgiger) [inline types/abeeway/badgeTracker/acknowledge.schema.json:2]: why does this need a separate schema?

File: types/abeeway/badgeTracker/activity_status.schema.json [hunk 2]
```diff
@@ -0,0 +1,13 @@
+{
+  "$id": "https://akenza.io/abeeway/compactTracker/activity_status.schema.json",
```
Comment (mariusgiger) [inline types/abeeway/badgeTracker/activity_status.schema.json:2]: why does this need a separate schema?

File: types/abeeway/badgeTracker/ble_beacon_failure.schema.json [hunk 3]
```diff
@@ -0,0 +1,22 @@
+{
+  "$id": "https://akenza.io/abeeway/compactTracker/ble_beacon_failure.schema.json",
```
Comment (mariusgiger) [inline types/abeeway/badgeTracker/ble_beacon_failure.schema.json:2]: why does this need a separate schema?

File: types/abeeway/badgeTracker/ble_beacon_long.schema.json [hunk 4]
```diff
@@ -0,0 +1,25 @@
+{
+  "$id": "https://akenza.io/abeeway/compactTracker/ble_beacon_long.schema.json",
```
Comment (mariusgiger) [inline types/abeeway/badgeTracker/ble_beacon_long.schema.json:2]: why does this need a separate schema? just add all these values to lifecycle

File: types/abeeway/badgeTracker/ble_connected.schema.json [hunk 5]
```diff
@@ -0,0 +1,13 @@
+{
+  "$id": "https://akenza.io/abeeway/compactTracker/ble_connected.schema.json",
```
Comment (mariusgiger) [inline types/abeeway/badgeTracker/ble_connected.schema.json:2]: placeholder?

File: types/abeeway/badgeTracker/ble_bond_deleted.schema.json [hunk 6]
```diff
@@ -0,0 +1,13 @@
+{
+  "$id": "https://akenza.io/abeeway/compactTracker/ble_bond_deleted.schema.json",
```
Comment (mariusgiger) [inline types/abeeway/badgeTracker/ble_bond_deleted.schema.json:2]: placeholder?

File: types/abeeway/badgeTracker/ble_disconnected.schema.json [hunk 7]
```diff
@@ -0,0 +1,13 @@
+{
+  "$id": "https://akenza.io/abeeway/compactTracker/ble_disconnected.schema.json",
```
Comment (mariusgiger) [inline types/abeeway/badgeTracker/ble_disconnected.schema.json:2]: placeholder?

File: types/abeeway/badgeTracker/encrypted_wifi_bssid.schema.json [hunk 8]
```diff
@@ -0,0 +1,13 @@
+{
+  "$id": "https://akenza.io/abeeway/compactTracker/encrypted_wifi_bssid.schema.json",
```
Comment (mariusgiger) [inline types/abeeway/badgeTracker/encrypted_wifi_bssid.schema.json:2]: placeholder?

File: types/abeeway/badgeTracker/meta.json [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+  "version": "1.0.0",
+  "manufacturer": "Abeeway",
+  "url": "https://www.abeeway.com/smart-badge/",
+  "description": "A multi-mode tracker with embedded sensors combining multi-constellation GNSS, Low-power GPS, Wi-Fi Sniffer, BLE and LoRaWAN®TDoA geolocation technologies, for seamless outdoor and indoor geolocation.",
+  "author": "Akenza AG",
+  "firmwareVersion": "V2.1.6",
+  "loraDeviceClass": "A",
+  "availableSensors": ["GPS"],
+  "outputTopics": [
+    "acknowledge",
```
Comment (mariusgiger) [inline types/abeeway/badgeTracker/meta.json:12]: I think this is definitely too many topics. Can you try to separate the messages into control and data messages and for the control ones you use the lifecycle topic?

File: types/abeeway/badgeTracker/uplink.js [hunk 10]
```diff
@@ -0,0 +1,981 @@
...
+  // Header
+  const type = Bits.bitsToUnsigned(bits.substr(0, 8));
+  lifecycle.demandMessage = !!Bits.bitsToUnsigned(bits.substr(15, 1));
+  lifecycle.positionMessage = !!Bits.bitsToUnsigned(bits.substr(14, 1));
+  lifecycle.hasMoved = !!Bits.bitsToUnsigned(bits.substr(13, 1));
+  // Reserved
+  lifecycle.sos = !!Bits.bitsToUnsigned(bits.substr(11, 1));
+  const operatingMode = Bits.bitsToUnsigned(bits.substr(8, 3));
+
+  switch (operatingMode) {
```
Comment (mariusgiger) [inline types/abeeway/badgeTracker/uplink.js:153]: possibly extract this into a separate function

File: types/abeeway/badgeTracker/uplink.js [hunk 11]
```diff
@@ -0,0 +1,981 @@
...
+      lifecycle.operatingMode = "ACTIVITY_TRACKING";
+      break;
+    case 5:
+      lifecycle.operatingMode = "OFF";
+      break;
+    default:
+      break;
+  }
+
+  const batteryLevel = Bits.bitsToUnsigned(bits.substr(16, 8));
```
Comment (mariusgiger) [inline types/abeeway/badgeTracker/uplink.js:176]: possibly extract this into a separate function

File: types/abeeway/badgeTracker/uplink.js [hunk 12]
```diff
@@ -0,0 +1,981 @@
...
+  } else if (batteryLevel === 255) {
+    lifecycle.batteryStatus = "ERROR";
+  } else {
+    lifecycle.batteryLevel = batteryLevel;
+  }
+  // prettier-ignore
+  lifecycle.temperature = Math.round(valueDecode(Bits.bitsToUnsigned(bits.substr(24, 8)), -44, 85, 8, 0) *10)/10;
+  // Reserved 4-8
+
+  switch (type) {
```
Comment (mariusgiger) [inline types/abeeway/badgeTracker/uplink.js:188]: possibly extract this into a separate function

File: types/abeeway/microTracker/uplink.js [hunk 13]
```diff
@@ -0,0 +1,981 @@
+function stepSize(lo, hi, nbits, nresv) {
```
Comment (mariusgiger) [inline types/abeeway/microTracker/uplink.js:1]: add a comment explaining that when changing this, the other types also need to be changed

**Review by mariusgiger — CHANGES_REQUESTED** [review]
nice work! I think we need to review the topic structure again, please prepare a page/diagram listing which message ends up in which topic.

File: types/abeeway/badgeTracker/activity_status.schema.json (hunk 2 above)
Comment (J2thatsme) [reply types/abeeway/badgeTracker/activity_status.schema.json:2]: activityCounter does only get sent here

File: types/abeeway/badgeTracker/gps_fix_extended.schema.json [hunk 14]
```diff
@@ -0,0 +1,54 @@
+{
+  "$id": "https://akenza.io/abeeway/compactTracker/gps_fix_extended.schema.json",
```
[the remark this answers is not in the document]
Comment (J2thatsme) [reply types/abeeway/badgeTracker/gps_fix_extended.schema.json]: @mariusgiger i cut a lot of the schemas. As indication which message is sent, i defined an enum. Please take a look, and tell me if it makes sense for you.

**Review by mariusgiger — APPROVED** [review]
Nice rework, let's see how it performs in real life.

**Review by mariusgiger — APPROVED** [review]

**Pull request merged**